pkg/ingester/ingester_v2.go

```diff
@@ -753,30 +741,19 @@ func (i *Ingester) v2Push(ctx context.Context, req *cortexpb.WriteRequest) (*cor
 			var err error
 
 			// If the cached reference exists, we try to use it.
-			if cachedRefExists {
-				var ref uint64
-				if ref, err = app.Append(cachedRef, copiedLabels, s.TimestampMs, s.Value); err == nil {
+			if ref != 0 {
+				labels := cortexpb.FromLabelAdaptersToLabels(ts.Labels)
+				if _, err = app.Append(ref, labels, s.TimestampMs, s.Value); err == nil {
 					succeededSamplesCount++
-					// This means the reference changes which means we need to update our cache.
-					if ref != cachedRef {
-						db.refCache.SetRef(startAppend, copiedLabels, ref)
-					}
 					continue
 				}
 
 			} else {
-				var ref uint64
-
 				// Copy the label set because both TSDB and the cache may retain it.
 				copiedLabels = cortexpb.FromLabelAdaptersToLabelsWithCopy(ts.Labels)
```

Comment:
Should be done only if `copiedLabels` is `nil` (think about the case of multiple samples for the same series in the same request).

Reply:
next time round `ref` will be non-zero so we won't execute this line.

Reply:
Oh right! 👍